File: app/src/main/java/com/porarri/yamabikochat/data/modelsdev/ModelsDevCatalog.kt
package com.porarri.yamabikochat.data.modelsdev

import kotlinx.serialization.Serializable

const val MODELS_DEV_PROVIDER_PREFIX = "MODELS_DEV:"

@Serializable
data class ProviderReference(
    val persistedId: String
) {
    val modelsDevId: String?
        get() = persistedId
            .takeIf { it.startsWith(MODELS_DEV_PROVIDER_PREFIX, ignoreCase = true) }
            ?.substringAfter(':')
            ?.trim()
            ?.lowercase()
            ?.takeIf { it.isNotEmpty() }

    val isModelsDev: Boolean get() = modelsDevId != null

    companion object {
        fun modelsDev(providerId: String): ProviderReference =
            ProviderReference("$MODELS_DEV_PROVIDER_PREFIX${providerId.trim().lowercase()}")
    }
}

object ModelsDevMergedProvider {
    fun catalogIdFor(persistedId: String): String? = ProviderReference(persistedId).modelsDevId
        ?: mapOf(
            "GEMINI" to "google", "OPENAI" to "openai"
        )[persistedId.uppercase()]
}

@Serializable
data class CatalogCost(
    val inputPerMillion: Double? = null,
    val outputPerMillion: Double? = null,
    val reasoningPerMillion: Double? = null,
    val cacheReadPerMillion: Double? = null,
    val cacheWritePerMillion: Double? = null
)

@Serializable
data class CatalogLimits(
    val context: Long? = null,
    val input: Long? = null,
    val output: Long? = null
)

@Serializable
data class CatalogReasoningOption(
    val type: String,
    val values: List<String> = emptyList()
)

@Serializable
data class CatalogModelProviderContract(
    val npm: String? = null,
    val api: String? = null,
    val shape: String? = null,
    val provenance: String? = null
)

object ModelsDevReasoningPreference {
    private const val FIELD_PREFIX = "YAMABIKO_REASONING_EFFORT_"

    fun fieldName(modelId: String): String = FIELD_PREFIX + modelId
        .toByteArray(Charsets.UTF_8)
        .joinToString(separator = "") { byte -> "%02X".format(byte.toInt() and 0xFF) }

    fun fieldKey(providerId: String, fieldName: String): String {
        val provider = providerId.lowercase().replace(Regex("[^a-z0-9._-]+"), "_")
        val field = fieldName.uppercase().replace(Regex("[^A-Z0-9_]+"), "_")
        return "models_dev_${provider}_$field"
    }

    fun storageKey(providerId: String, modelId: String): String =
        fieldKey(providerId, fieldName(modelId))
}

@Serializable
data class CatalogModel(
    val id: String,
    val name: String,
    val description: String? = null,
    val family: String? = null,
    val attachment: Boolean? = null,
    val reasoning: Boolean? = null,
    val reasoningOptions: List<CatalogReasoningOption> = emptyList(),
    val toolCall: Boolean? = null,
    val structuredOutput: Boolean? = null,
    val temperature: Boolean? = null,
    val inputModalities: List<String> = emptyList(),
    val outputModalities: List<String> = emptyList(),
    val releaseDate: String? = null,
    val lastUpdated: String? = null,
    val limits: CatalogLimits = CatalogLimits(),
    val cost: CatalogCost = CatalogCost(),
    val providerContract: CatalogModelProviderContract? = null
) {
    val supportedReasoningEfforts: List<String>
        get() = reasoningOptions
            .asSequence()
            .filter { it.type.equals("effort", ignoreCase = true) }
            .flatMap { it.values.asSequence() }
            .map { it.trim().lowercase() }
            .filter { it.isNotEmpty() }
            .distinct()
            .toList()

    fun shouldShowReasoningEffortPreference(savedEffort: String): Boolean =
        supportedReasoningEfforts.isNotEmpty() || savedEffort.isNotBlank()

    fun matches(query: String): Boolean {
        val normalized = query.trim().lowercase()
        if (normalized.isEmpty()) return true
        return listOf(id, name, family.orEmpty(), description.orEmpty())
            .any { it.lowercase().contains(normalized) }
    }
}

@Serializable
data class CatalogProvider(
    val id: String,
    val name: String,
    val npm: String,
    val api: String? = null,
    val env: List<String> = emptyList(),
    val documentationUrl: String? = null,
    val models: List<CatalogModel> = emptyList()
) {
    val reference: ProviderReference get() = ProviderReference.modelsDev(id)

    fun matches(query: String): Boolean {
        val normalized = query.trim().lowercase()
        return normalized.isEmpty() || id.lowercase().contains(normalized) || name.lowercase().contains(normalized)
    }
}

enum class CatalogAvailability { IDLE, LOADING, READY, STALE, ERROR }

data class CatalogLoadState(
    val availability: CatalogAvailability = CatalogAvailability.IDLE,
    val providers: List<CatalogProvider> = emptyList(),
    val lastUpdatedEpochMs: Long? = null,
    val error: String? = null
)

File: ios/YamabikoChat/Data/Models/ModelsDevCatalog.swift
import Foundation

let modelsDevProviderPrefix = "MODELS_DEV:"

struct ProviderReference: Codable, Hashable, Sendable {
    let persistedID: String

    var modelsDevID: String? {
        guard persistedID.uppercased().hasPrefix(modelsDevProviderPrefix) else { return nil }
        let value = String(persistedID.dropFirst(modelsDevProviderPrefix.count))
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        return value.isEmpty ? nil : value
    }

    var isModelsDev: Bool { modelsDevID != nil }

    static func modelsDev(_ providerID: String) -> ProviderReference {
        ProviderReference(persistedID: modelsDevProviderPrefix + providerID.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())
    }
}

enum ModelsDevMergedProvider {
    private static let catalogProviderIDsRepresentedByBuiltIns: Set<String> = ["google", "openrouter"]

    static func catalogID(for persistedID: String) -> String? {
        let reference = ProviderReference(persistedID: persistedID)
        if let dynamic = reference.modelsDevID { return dynamic }
        return persistedID.caseInsensitiveCompare("GEMINI") == .orderedSame ? "google" : nil
    }

    static func isSelectableCatalogProvider(_ providerID: String) -> Bool {
        !catalogProviderIDsRepresentedByBuiltIns.contains(providerID.lowercased())
    }
}

struct CatalogReasoningOption: Codable, Equatable, Sendable {
    let type: String
    var values: [String] = []
}

enum ModelsDevReasoningPreference {
    private static let fieldPrefix = "YAMABIKO_REASONING_EFFORT_"

    static func fieldName(modelID: String) -> String {
        let encodedModelID = modelID.utf8.map { String(format: "%02X", $0) }.joined()
        return fieldPrefix + encodedModelID
    }

    static func fieldKey(providerID: String, fieldName: String) -> String {
        let provider = providerID.lowercased().replacingOccurrences(
            of: "[^a-z0-9._-]+",
            with: "_",
            options: .regularExpression
        )
        let field = fieldName.uppercased().replacingOccurrences(
            of: "[^A-Z0-9_]+",
            with: "_",
            options: .regularExpression
        )
        return "models_dev_\(provider)_\(field)"
    }

    static func storageKey(providerID: String, modelID: String) -> String {
        fieldKey(providerID: providerID, fieldName: fieldName(modelID: modelID))
    }
}

struct CatalogLimits: Codable, Equatable, Sendable {
    var context: Int?
    var input: Int?
    var output: Int?
}

struct CatalogCost: Codable, Equatable, Sendable {
    var inputPerMillion: Double?
    var outputPerMillion: Double?
    var reasoningPerMillion: Double?
    var cacheReadPerMillion: Double?
    var cacheWritePerMillion: Double?
}

struct CatalogModelProviderContract: Codable, Equatable, Sendable {
    var npm: String?
    var api: String?
    var shape: String?
}

struct CatalogModel: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let name: String
    var description: String?
    var family: String?
    var attachment: Bool?
    var reasoning: Bool?
    var reasoningOptions: [CatalogReasoningOption]
    var toolCall: Bool?
    var structuredOutput: Bool?
    var temperature: Bool?
    var inputModalities: [String]
    var outputModalities: [String]
    var releaseDate: String?
    var lastUpdated: String?
    var limits: CatalogLimits
    var cost: CatalogCost
    var providerContract: CatalogModelProviderContract? = nil

    var supportedReasoningEfforts: [String] {
        var seen = Set<String>()
        return reasoningOptions
            .filter { $0.type.caseInsensitiveCompare("effort") == .orderedSame }
            .flatMap(\.values)
            .compactMap { rawValue in
                let value = rawValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                guard !value.isEmpty, seen.insert(value).inserted else { return nil }
                return value
            }
    }

    func shouldShowReasoningEffortPreference(savedEffort: String) -> Bool {
        !supportedReasoningEfforts.isEmpty || !savedEffort.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func matches(_ query: String) -> Bool {
        let value = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return true }
        return [id, name, family ?? "", description ?? ""].contains {
            $0.localizedCaseInsensitiveContains(value)
        }
    }
}

struct CatalogProvider: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let name: String
    let npm: String
    var api: String?
    var env: [String]
    var documentationURL: String?
    var models: [CatalogModel]

    var reference: ProviderReference { .modelsDev(id) }

    func matches(_ query: String) -> Bool {
        let value = query.trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty || id.localizedCaseInsensitiveContains(value) || name.localizedCaseInsensitiveContains(value)
    }
}

enum CatalogAvailability: String, Codable, Sendable {
    case idle, loading, ready, stale, error
}

struct CatalogLoadState: Equatable, Sendable {
    var availability: CatalogAvailability = .idle
    var providers: [CatalogProvider] = []
    var lastUpdated: Date?
    var error: String?
}
